cache/jsonl.ts

```diff
@@ -0,0 +1,241 @@
+      } catch {
+        return false;
+      }
+    });
+  }
+
+  /**
+   * Returns content of a file as a stream
+   *
+   * ```ts
+   * import { each } from "npm:effection@4.0.0-alpha.3";
+   * import { useCache } from "jsr:@effection-contrib/cache";
+   *
+   * const cache = yield* useCache();
+   *
+   * for (const item of yield* each(yield* cache.read<number>("test"))) {
+   *   console.log(item)
+   *   yield* each.next();
+   * }
+   * ```
+   *
+   * @param key string
+   * @returns Stream<T>
+   */
+  *read<T>(key: string): Operation<Stream<T, void>> {
+    const location = new URL(`./${key}.jsonl`, this.location);
+    const file = yield* call(() => Deno.open(location, { read: true }));
+
+    const lines = file
+      .readable
+      .pipeThrough(new TextDecoderStream())
+      .pipeThrough(new JSONLinesParseStream());
+
+    return stream(lines as ReadableStream<T>);
+  }
+
+  /**
+   * Write data to a file, creates the file and necessary directory structure as it goes along.
+   *
+   * ```ts
+   * import { useCache } from "jsr:@effection-contrib/cache";
+   *
+   * const cache = yield* useCache();
+   * yield* cache.write("hello", "world");
+   * ```
+   * @param key string
+   * @param data unknown
+   */
+  *write(key: string, data: unknown): Operation<void> {
+    const location = new URL(`./${key}.jsonl`, this.location);
+
+    yield* mkdir(dirname(location.pathname), { recursive: true });
+
+    const file = yield* call(() =>
+      Deno.open(location, {
+        create: true,
+        write: true,
+      })
+    );
+
+    try {
+      yield* call(() =>
+        file.write(new TextEncoder().encode(`${JSON.stringify(data)}\n`))
+      );
+    } finally {
+      file.close();
+    }
+  }
+
+  /**
+   * Add data to an existing file.
+   *
+   * ```ts
+   * import { useCache } from "jsr:@effection-contrib/cache";
+   *
+   * const cache = yield* useCache();
+   * yield* cache.write("hello", "world");
+   * yield* cache.append("hello", "from bob");
+   * ```
+   * @param key string
+   * @param data
+   */
+  *append(key: string, data: unknown): Operation<void> {
+    const location = new URL(`./${key}.jsonl`, this.location);
+
+    const file = yield* call(() =>
+      Deno.open(location, {
+        append: true,
+      })
+    );
+
+    try {
+      yield* call(() =>
+        file.write(new TextEncoder().encode(`${JSON.stringify(data)}\n`))
+      );
+    } finally {
+      file.close();
+    }
+  }
+
+  /**
+   * Returns a stream of content from all files matching a glob
+   *
+   * ```ts
+   * import { each } from "npm:effection@4.0.0-alpha.3";
+   * import { useCache } from "jsr:@effection-contrib/cache";
+   *
+   * const cache = yield* useCache();
+   *
+   * for (const item of yield* each(cache.find<number>("subdir/*"))) {
+   *   console.log(item);
+   *    yield* each.next();
+   * }
+   * ```
+   *
+   * @param glob string
+   * @returns Stream<T, void>
+   */
+  *find<T>(glob: string): Stream<T, void> {
```

Comment:
You may want to consider making this a resource:

```ts
find<T>(glob: string): Stream<T, void> {
  return resource(function*(provide) {
     // setup and provide queue
  });
}
```

It will function exactly the same, except in the inspector (which doesn't exist yet) all of the children will be identified as being part of the `cache.find()` resource, and not "inlined" into the caller of `cache.find()`

Reply:
Would this be the same for read?

Reply:
Yes, anywhere you are creating a stream for yourself.